# Supply Chain Analytics in Tableau or Power BI

## 📖 Background
Test your BI skills on a real-world dataset focusing on supply chain analytics. As the main data analyst for Just In Time, you will help solve key shipment and inventory management challenges, analyze supply chain inefficiencies, and create insightful dashboards to inform business stakeholders about potential problems and propose structural business improvements.

Be creative and make use of your full skillset! Use this Workspace to prepare your data, import the tables into your local Tableau or Power BI instance, and share your insights below.

The end goal will be a (set of) interactive dashboards that demonstrate clear insights for Just In Time.

## 💾 The data


| Group | Column name | Dataset | Definition |
|:---|:---|:---|:---|
| Customer | Customer ID | orders_and_shipments.csv | Unique customer identification |
| Customer | Customer Market | orders_and_shipments.csv | Geographic grouping of customer countries, with values such as Europe, LATAM, Pacific Asia, etc. |
| Customer | Customer Region | orders_and_shipments.csv | Geographic grouping of customer countries, with values such as Northern Europe, Western Europe, etc. |
| Customer | Customer Country | orders_and_shipments.csv | Customer's country |
| Order info | Order ID | orders_and_shipments.csv | Unique Order identification. Order groups one or multiple Order Items |
| Order info | Order Item ID | orders_and_shipments.csv | Unique Order Item identification. Order Item always belong to just one Order |
| Order info | Order Year | orders_and_shipments.csv | Year of the order |
| Order information | Order Month | orders_and_shipments.csv | Month of the order |
| Order information | Order Day | orders_and_shipments.csv | Day of the order |
| Order information | Order Time | orders_and_shipments.csv | Timestamp of the order in UTC |
| Order information | Order Quantity | orders_and_shipments.csv | The amount of items that were ordered within a given Order Item (1 record of the data) |
| Product | Product Department | orders_and_shipments.csv | Product grouping into categories such as Fitness, Golf, Pet Shop, etc. |
| Product | Product Category | orders_and_shipments.csv | Product grouping into categories such as Sporting Goods, Women's Apparel, etc. |
| Product | Product Name | orders_and_shipments.csv | The name of the purchased product |
| Sales | Gross Sales | orders_and_shipments.csv | Revenue before discounts generated by the sales of the Order Item (1 record of the data) |
| Sales | Discount % | orders_and_shipments.csv | Discount % applied on the catalog price |
| Sales | Profit | orders_and_shipments.csv | Profit generated by the sales of the Order Item (1 record of data) |
| Shipment information | Shipment Year | orders_and_shipments.csv | Year of the shipment |
| Shipment information | Shipment Month | orders_and_shipments.csv | Month of the shipment |
| Shipment information | Shipment Day | orders_and_shipments.csv | Day of the shipment |
| Shipment information | Shipment Mode | orders_and_shipments.csv | Information on how the shipment has been dispatched, with values as First Class, Same Day, Second Class, etc. |
| Shipment information | Shipment Days - Scheduled | orders_and_shipments.csv | Information on typical amount of days needed to dispatch the goods from the moment the order has been placed |
| Warehouse | Warehouse Country | orders_and_shipments.csv | Country of the warehouse that has fulfilled this order, the only two values being Puerto Rico and USA |
| Inventory & Fulfillment | Warehouse Inventory | inventory.csv | The monthly level of inventory of a product, e.g. 930 units |
| Inventory & Fulfillment | Inventory cost per unit | inventory.csv | The monthly storage cost per unit of inventory, e.g. $2.07 |
| Inventory & Fulfillment | Warehouse Order fulfillment (days) | fulfillment.csv | The average amount of days it takes to refill stock if inventory drops below zero |


**The data can be downloaded from the sidebar on the left (under _Files_).**

## 💪 Challenge
Using either Tableau or Power BI, create an interactive dashboard to summarize your research. Things to consider:

1. Use this Workspace to prepare your data (optional).
2. Some ideas to get you started: visualize how shipments are delayed, by country, product, and over time. Analyze products by their supply versus demand ratio. Rank products by over or understock. Don't feel limited by these, you're encouraged to use your skills to consolidate as much information as possible. 
3. Create a screenshot of your (main) Tableau or Power BI dashboard, and paste that into the designated field.
4. Summarize your findings in an executive summary.

In [1]:
import pandas as pd
data = pd.read_csv("data/orders_and_shipments.csv")
data.head(3)

Order ID    Order Item ID    Order YearMonth    Order Year    Order Month   \
0       3535             8793             201502          2015              2   
1       4133            10320             201503          2015              3   
2       7396            18517             201504          2015              4   

    Order Day  Order Time  Order Quantity Product Department Product Category  \
0           21      14:07               1           Fan Shop          Fishing   
1            2      07:37               1           Fan Shop          Fishing   
2           18      22:47               1           Fan Shop          Fishing   

   ... Customer Country  Warehouse Country Shipment Year Shipment Month  \
0  ...           Mexico        Puerto Rico          2015              2   
1  ...           Brazil        Puerto Rico          2015              3   
2  ...           Mexico        Puerto Rico          2015              4   

  Shipment Day   Shipment Mode   Shipment Days - Scheduled    Gross Sales   \
0           27  Standard Class                            4            400   
1            6  Standard Class                            4            400   
2           20  Standard Class                            4            400   

    Discount %   Profit   
0          0.25      200  
1          0.09      200  
2          0.06      200  

[3 rows x 24 columns]

In [2]:
# check the column names
print(data.columns)

Index(['Order ID ', ' Order Item ID ', ' Order YearMonth ', ' Order Year ',
       ' Order Month ', ' Order Day ', 'Order Time', 'Order Quantity',
       'Product Department', 'Product Category', 'Product Name',
       ' Customer ID ', 'Customer Market', 'Customer Region',
       'Customer Country', 'Warehouse Country', 'Shipment Year',
       'Shipment Month', 'Shipment Day', 'Shipment Mode',
       ' Shipment Days - Scheduled ', ' Gross Sales ', ' Discount % ',
       ' Profit '],
      dtype='object')


In [3]:
# remove extra spaces from column names
data.columns = data.columns.str.strip()
print(data.columns)

Index(['Order ID', 'Order Item ID', 'Order YearMonth', 'Order Year',
       'Order Month', 'Order Day', 'Order Time', 'Order Quantity',
       'Product Department', 'Product Category', 'Product Name', 'Customer ID',
       'Customer Market', 'Customer Region', 'Customer Country',
       'Warehouse Country', 'Shipment Year', 'Shipment Month', 'Shipment Day',
       'Shipment Mode', 'Shipment Days - Scheduled', 'Gross Sales',
       'Discount %', 'Profit'],
      dtype='object')


In [4]:
def create_date_columns(df, day_col, month_col, year_col, col_name):
    """
    Create a new column with a datetime object from three columns with day, month and year.
    """
    df[col_name] = pd.to_datetime(df[month_col].astype(str) + '-' + df[day_col].astype(str) + '-' + df[year_col].astype(str))

In [5]:
# create order date column
create_date_columns(data, 'Order Day', 'Order Month', 'Order Year', 'Order Date')

# create shipment date column
create_date_columns(data, 'Shipment Day', 'Shipment Month', 'Shipment Year', 'Shipment Date')

In [6]:
def create_datetime_columns(df, day_col, month_col, year_col, time_col, col_name):
    """
    Create a new column with a datetime object from three columns with day, month and year.
    """
    df[col_name] = pd.to_datetime(df[month_col].astype(str) + '-' + df[day_col].astype(str) + '-' + df[year_col].astype(str) + ' ' + df[time_col].astype(str))

In [7]:
# create order datetime column
create_datetime_columns(data, 'Order Day', 'Order Month', 'Order Year', 'Order Time', 'Order Datetime')

In [8]:
# create shipment days column [actual time to place shipment]
data['Shipment Days'] = (data['Shipment Date'] - data['Order Date']).dt.days

In [9]:
import numpy as np

# create shipment delay column
data['Shipment Delay'] = np.where(data['Shipment Days - Scheduled'] > data['Shipment Days'], 0, data['Shipment Days'] - data['Shipment Days - Scheduled'])

In [10]:
# if shipment delay is negative there might be error in the shipment month
fil_df = np.where(data['Shipment Days'] < 0 )

In [12]:
filtered_shipment_date = data.loc[fil_df, 'Shipment Date']
filtered_order_date = data.loc[fil_df, 'Order Date']
filtered_shipment_day = data.loc[fil_df, 'Shipment Day']
filtered_shipment_month = data.loc[fil_df, 'Shipment Month']
filtered_shipment_year = data.loc[fil_df, 'Shipment Year']
filtered_order_day = data.loc[fil_df, 'Order Day']
filtered_order_month = data.loc[fil_df, 'Order Month']
filtered_order_year = data.loc[fil_df, 'Order Year']

# Create a new DataFrame to combine the 'Shipment Date' and 'Order Date'
combined_data = pd.DataFrame({'Order Day': filtered_order_day, 'Order Month': filtered_order_month, 'Order Year': filtered_order_year, 
                              'Order Date': filtered_order_date, 
                              'Shipment Day': filtered_shipment_day, 'Shipment Month': filtered_shipment_month, 'Shipment Year': filtered_shipment_year,
                              'Shipment Date': filtered_shipment_date, })

# Print the combined DataFrame
combined_data

Order Day  Order Month  Order Year Order Date  Shipment Day  \
294           10            2        2015 2015-02-10            20   
295            2            3        2015 2015-03-02            14   
296            4            4        2015 2015-04-04            17   
297            6            4        2015 2015-04-06            23   
298           18            9        2015 2015-09-18            19   
...          ...          ...         ...        ...           ...   
30560         25            5        2017 2017-05-25            16   
30561         13            7        2017 2017-07-13            12   
30562         21            7        2017 2017-07-21            13   
30563         10            8        2017 2017-08-10            21   
30564          9            9        2017 2017-09-09            18   

       Shipment Month  Shipment Year Shipment Date  
294                 1           2015    2015-01-20  
295                 2           2015    2015-02-14  
296                 3           2015    2015-03-17  
297                 3           2015    2015-03-23  
298                 7           2015    2015-07-19  
...               ...            ...           ...  
30560               6           2016    2016-06-16  
30561               7           2016    2016-07-12  
30562               7           2016    2016-07-13  
30563               7           2016    2016-07-21  
30564               7           2016    2016-07-18  

[2735 rows x 8 columns]

In [28]:
wrong_indices = np.where(data['Shipment Date'] < data['Order Date'])
data.loc[wrong_indices, 'Shipment Date'].count()

# Remove the wrong rows from the DataFrame
mask = np.where(data['Shipment Date'] >= data['Order Date'])
data = data.loc[mask]

In [30]:
data.shape[0]

28136

So there are 2735 orders which shipment dates preceed the order dates which is not possible in a real life scenario. So, we have to remove those from our data when we perform data validation to maintain the data quality.

In [14]:
# check the customer country column values
data['Customer Country'].unique()


array(['Mexico', 'Brazil', 'Denmark', 'Netherlands', 'Germany', 'China',
       'Indonesia', 'Pakistan', 'India', 'USA', 'Hungary', 'Sudan',
       'Democratic Republic of Congo', 'Poland', 'Togo', 'Guatemala',
       'Panama', 'Chile', 'France', 'Sweden', 'Dominican�Republic',
       'Venezuela', 'South Korea', 'Madagascar', 'Iran', 'Cuba',
       'Nicaragua', 'United Kingdom', 'Afghanistan', 'Singapore',
       'Morocco', 'Spain', 'Niger', 'Turkey', 'South Africa', 'Iraq',
       'Honduras', 'Italy', 'Australia', 'Cote d�Ivoire', 'Croatia',
       'Ecuador', 'Syria', 'Haiti', 'Bangladesh', 'Argentina', 'Romania',
       'El Salvador', 'Vietnam', 'Japan', 'Nigeria', 'Belarus',
       'Uzbekistan', 'Egypt', 'Albania', 'Georgia', 'Cameroon',
       'Colombia', 'New zealand', 'Canada', 'Thailand', 'Senegal',
       'Russia', 'Per�', 'Algeria�', 'Ukraine', 'Belgium', 'Philippines',
       'Austria', 'Uruguay', 'Malaysia', 'Hong Kong', 'Saudi Arabia',
       'Switzerland', 'Ireland', 'Bulg

In [15]:
data['Customer Country'] = np.where(data['Customer Country'] == 'Cote d�Ivoire', "Cote d'Ivoire", data['Customer Country'])
data['Customer Country'] = np.where(data['Customer Country'] == 'Dominican�Republic', 'Dominican Republic', data['Customer Country'])
data['Customer Country'] = np.where(data['Customer Country'] == 'Per�', 'Peru', data['Customer Country'])
data['Customer Country'] = np.where(data['Customer Country'] == 'Algeria�', 'Algeria', data['Customer Country'])
data['Customer Country'] = np.where(data['Customer Country'] == 'Israel�', 'Israel', data['Customer Country'])
data['Customer Country'] = np.where(data['Customer Country'] == 'Ben�n', 'Benin', data['Customer Country'])

In [16]:
# Assuming 'data' is a pandas DataFrame with 'Gross Sales' and 'Discount %' columns
# First, convert 'Discount %' to a numeric type (if it's not already)
data['Discount %'] = pd.to_numeric(data['Discount %'], errors='coerce')

# Calculate the discounted price for each row
data['Net Sales'] = data['Gross Sales'] - (data['Gross Sales'] * (data['Discount %'] / 100))

In [32]:
data.to_csv('data/orders_and_shipments_transformed.csv', index=False)

In [18]:
# import pandas as pd
# data = pd.read_csv("data/orders_and_shipments_transformed.csv")

## ✅ Checklist before publishing
- If you use Tableau, don't forget to publish your Tableau dashboard, make it available on Tableau Public and share the link.
- If you use Power BI, upload your `.pbix` file to this Workspace through the sidebar on the left (under _Files_).
- **Remove redundant text cells** like the background, data, challenge, and checklist. You can add cells if necessary.

## ✍️ Judging criteria
| CATEGORY | WEIGHTING | DETAILS                                                              |
|:---------|:----------|:---------------------------------------------------------------------|
| **Visualizations** | 35% | <ul><li>Appropriateness of visualizations used.</li><li>Clarity of insight from visualizations.</li></ul> |
| **Insights** | 25%       | <ul><li>Clarity of insights - how clear and well presented the insights are.</li><li>Quality of recommendations - are appropriate analytical techniques used & are the conclusions valid?</li><li>Number of relevant insights found for the target audience.</li></ul>       |
| **Storytelling**  | 25%       | <ul><li>How well the data and insights are connected to tell a story.</li><li>How the narrative and whole report connects together.</li><li>How balanced the report is: in-depth enough but also concise.</li></ul> |
| **Votes** | 15% | <ul><li>Up voting - most upvoted entries get the most points.</li></ul> |

## 🧾 Executive summary
_In a couple of lines, write your main findings here._

## 📷 Dashboard screenshot
_Paste one screenshot of your Tableau or Power BI dashboard here._

## 🌐 Upload your dashboard
For Tableau: paste the link to your Tableau Public dashboard here.

For Power BI: upload your `.pbix` file to _Files_ in the sidebar on the left.



## ⌛️ Time is ticking. Good luck!
